File: tests/pure/test_stale_obnovit.rb
# frozen_string_literal: true
# Jantarovy indikator neaktualnosti tlacidla „Obnoviť" v okne ŠTÚDIO (22.8.).
#
# CO to riesi: Studio cisla neprepocitava samo — kym sa nestlaci „Obnoviť",
# visia v nom cisla z posledneho prepoctu. Model sa medzitym mohol zmenit
# (prestavba skrinky z Inspectora, posun, Spat/Znova) a okno vyzeralo uplne
# rovnako — dalo sa teda exportovat VEPO, objednavku aj cenovu ponuku zo
# starych cisel.
#
# Observer sa headless spustit neda (potrebuje zivy SketchUp) — plne scenare
# su v tests/sketchup/su_runner.rb (sekcia `run_stale`, pocitadla vstupov).
# Tu su STATICKE guardy na kontrakt, ktory sa nesmie stratit pri buducej
# uprave studio_dialog.rb / studio.js / budget.js.
require_relative '../helper' unless defined?(NxTest)

module NxStale
  RB = File.read(File.join(NxTest::ROOT, 'noxun_engine', 'ui', 'studio_dialog.rb'), encoding: 'UTF-8')
  JS = File.read(File.join(NxTest::ROOT, 'noxun_engine', 'ui', 'js', 'studio.js'), encoding: 'UTF-8')
  BUD = File.read(File.join(NxTest::ROOT, 'noxun_engine', 'ui', 'js', 'budget.js'), encoding: 'UTF-8')
  HTML = File.read(File.join(NxTest::ROOT, 'noxun_engine', 'ui', 'studio.html'), encoding: 'UTF-8')
  MAIN = File.read(File.join(NxTest::ROOT, 'noxun_engine', 'main.rb'), encoding: 'UTF-8')

  def self.body(src, sig)
    src[/#{Regexp.escape(sig)}.*?\n        end\n/m].to_s
  end
end

# --- 1) observer: 4 hooky a NIC INE ------------------------------------------

NxTest.test('STALE: StudioModelWatch pokryva commit, undo, redo aj abort') do
  NxTest.assert(NxStale::RB.include?('class StudioModelWatch < Sketchup::ModelObserver'),
                'observer chyba — okno by o zmenach v modeli vobec nevedelo')
  # Bez guardu by tento subor v headless testoch spadol uz pri nacitani
  # (Sketchup tam neexistuje) — vzor core/edge_overlay.rb.
  NxTest.assert(NxStale::RB.include?('if defined?(Sketchup::ModelObserver)'),
                'trieda je pod guardom dostupnosti SketchUp API')
  %w[onTransactionCommit onTransactionUndo onTransactionRedo onTransactionAbort].each do |cb|
    NxTest.assert(NxStale::RB.include?("def #{cb}(model)"), "chyba callback #{cb}")
  end
  # PanelModelObserver pocuje LEN Spat/Znova/Abort a ScaleWatch vedome filtruje
  # vlastne prestavby — hlavny pripad (prestavba skrinky z Inspectora) by teda
  # cez ne neprisiel. Preto ma okno VLASTNY observer a tie dva ostavaju nedotknute.
  klass = NxStale::RB[/class StudioModelWatch.*?\n        end\n      end/m].to_s
  NxTest.assert_equal(4, klass.scan('StudioDialog.on_model_txn(model)').length,
                      'kazdy hook konci v TOM ISTOM tenkom handleri')
end

NxTest.test('STALE: callback je PRAZDNY — ziadne citanie ani zapis modelu') do
  body = NxStale.body(NxStale::RB, 'def on_model_txn(model)')
  NxTest.refute(body.empty?, 'handler sa nasiel')
  NxTest.assert(body.include?('@epoch = @epoch.to_i + 1'), 'callback LEN zdvihne epochu')
  NxTest.assert(body.include?('request_stale_flush(model)'), 'a naplanuje odlozeny flush')
  # Lekcia D-103: co v observer kontexte siahne na model, stane sa vrcholom
  # undo stacku hned po pouzivatelovom kroku.
  %w[push_state fresh_collect start_operation commit_operation Store\. Bom\. collect
     ProductionCore dedup_copies set_attribute].each do |forbidden|
    NxTest.refute(body.match?(/#{forbidden}/),
                  "v callbacku sa NESMIE objavit #{forbidden} — observer nic nepocita ani nezapisuje")
  end
end

NxTest.test('STALE: latch zluci burst commitov do JEDNEHO js') do
  body = NxStale.body(NxStale::RB, 'def request_stale_flush(model)')
  NxTest.refute(body.empty?, 'latch sa nasiel')
  NxTest.assert(body.include?('return if @stale_pending'),
                'dalsia udalost pocas pending uz dalsi timer nepridava (vzor Panel.request_txn_refresh)')
  NxTest.assert(body.include?('UI.start_timer(0, false)'),
                'flush bezi az MIMO observer kontextu')
  # Review #6: zapadka sa zamyka AZ PO tom, co timer vznikol — inak by vynimka
  # z `UI.start_timer` nechala okno TICHE az do zatvorenia.
  NxTest.assert(body.index('UI.start_timer(0, false)') < body.index('@stale_pending = true'),
                'zapadka sa zamyka az za planovanim timera')
  NxTest.assert(body.match?(/rescue StandardError => e\n\s*@stale_pending = false/),
                'a pri zlyhani sa OTVARA spat (ziadny trvaly tichy vypadok)')
end

# --- 2) POROVNANIE EPOCH (jadro riesenia self-tickov) ------------------------

NxTest.test('STALE: flush posiela markStale LEN ked je epocha NOVSIA nez pushnuta') do
  body = NxStale.body(NxStale::RB, 'def flush_stale(model)')
  NxTest.refute(body.empty?, 'flush sa nasiel')
  NxTest.assert(body.include?('return unless txn_model_ok?(model)'),
                'guard dokumentu sa overuje ZNOVA tesne pred pushom')
  NxTest.assert(body.include?('return unless @dialog && @dialog.visible?'),
                'do zavreteho okna sa neposiela nic')
  NxTest.assert(body.include?('return unless @epoch.to_i > @pushed_epoch.to_i'),
                'JADRO: vlastne ticky okna (dedup kopii, zapis rozpoctu) sa pohltia samy')
  NxTest.assert(body.include?('NX.markStale'), 'a klientovi ide JEDINY jednoduchy signal')
  # Guard sa overuje DVAKRAT (v callbacku aj v flushi) — dokument sa moze
  # prepnut medzi udalostou a timerom.
  NxTest.assert_equal(2, NxStale::RB.scan(/return unless txn_model_ok\?\(model\)/).length)
  guard = NxStale.body(NxStale::RB, 'def txn_model_ok?(model)')
  NxTest.assert(guard.include?('@observer_model') && guard.include?('Sketchup.active_model'),
                "dvojity model guard chyba: #{guard}")
end

NxTest.test('STALE: push_state si epochu uklada AZ PO zbere a LEN po odoslani') do
  push = NxStale::RB[/def push_state\(bump: true\).*?\n        end\n/m].to_s
  NxTest.refute(push.empty?, 'push_state sa nasiel')
  NxTest.assert(push.include?('@pushed_epoch = @epoch.to_i if sent'),
                'epocha „co uz je v okne" sa zapisuje LEN po OVEREN0M odoslani payloadu')
  collect_at = push.index('ProductionCore.fresh_collect(model)')
  mark_at = push.index('@pushed_epoch = @epoch.to_i if sent')
  NxTest.assert(!collect_at.nil? && !mark_at.nil? && collect_at < mark_at,
                'zapis epochy je AZ ZA `fresh_collect` — inak by vlastny dedup kopii ' \
                'nechal tlacidlo vecne jantarove')
end

# --- 3) lifecycle: attach / detach / prepnutie dokumentu ---------------------

NxTest.test('STALE: observer zije presne tak dlho ako okno') do
  ens = NxStale.body(NxStale::RB, 'def ensure_dialog')
  NxTest.assert(ens.include?('attach_stale_observer(Sketchup.active_model)'),
                'otvorenie okna observer zavesi')
  closed = NxStale::RB[/@dialog\.set_on_closed do.*?\n          end\n/m].to_s
  NxTest.assert(closed.include?('detach_stale_observer'), 'zatvorenie okna ho odvesi')
  chg = NxStale.body(NxStale::RB, 'def on_model_changed(model)')
  # Review #1: PODANY model, nie `active_model` — pri zlom nacasovani broadcastu
  # by observer ostal visiet na starom dokumente a indikator by uz nikdy
  # nezozltol. `active_model` je tu len zaloha, ked broadcast model nenesie.
  NxTest.assert(chg.include?('attach_stale_observer(model || Sketchup.active_model)'),
                'prepnutie dokumentu observer PREVESI na PODANY model (epocha je per dokument)')
  # Druha poistka: okno, ktore pocita cisla z ineho modelu, nez na akom visi
  # observer, sa prevesi samo (zmeskany broadcast tak nie je trvaly vypadok).
  push = NxStale::RB[/def push_state\(bump: true\).*?\n        end\n/m].to_s
  NxTest.assert(push.include?('attach_stale_observer(model) if @observer_model && @observer_model != model'),
                'push_state ma SAMOLIECBU — prevesi observer na model, z ktoreho prave pocita')
  NxTest.assert(push.index('attach_stale_observer(model) if') < push.index('sent = js('),
                'a robi to PRED odoslanim payloadu (epocha sa nuluje spolu s prevesenim)')
  att = NxStale.body(NxStale::RB, 'def attach_stale_observer(model = nil)')
  NxTest.assert(att.include?('m.remove_observer(@stale_observer)') && att.include?('m.add_observer(@stale_observer)'),
                'anti-double: remove PRED add (vzor EdgeCheck.attach_observer)')
  NxTest.assert(att.index('m.remove_observer') < att.index('m.add_observer'),
                'a v tomto poradi — inak by udalost prisla dvakrat')
  NxTest.assert(att.include?('rescue StandardError'),
                'zlyhanie odvesenia nesmie preskocit zavesenie (kazdy krok vlastny rescue)')
  det = NxStale.body(NxStale::RB, 'def detach_stale_observer')
  NxTest.assert(det.include?('reset_stale_epoch'), 'detach nuluje epochu')
  # Review #5: odvesenie z uz zavreteho dokumentu je OCAKAVANY stav — do logu
  # nepatri (rovnako ako v attachi).
  NxTest.refute(det.include?('log_error'),
                'zlyhanie odvesenia sa hlta TICHO (nie je to chyba)')
  reset = NxStale.body(NxStale::RB, 'def reset_stale_epoch')
  NxTest.assert(reset.include?('@epoch = 0') && reset.include?('@pushed_epoch = 0'),
                'nuluju sa OBE — inak by nove okno zacalo s cudzim rozdielom')
end

NxTest.test('STALE: ziadny Engine broadcast — vsetko zije v StudioDialog') do
  # Vedome: epocha ma JEDNEHO vlastnika. Broadcast by znamenal druhy zdroj
  # pravdy o tom, ci su cisla v okne stare.
  %w[notify_model_dirty push_stale broadcast_stale].each do |sym|
    NxTest.refute(NxStale::MAIN.include?(sym), "main.rb nema vediet o `#{sym}`")
  end
end

# --- 4) klient: JEDEN markup na 5 miestach ----------------------------------

NxTest.test('STALE: tlacidlo „Obnoviť" ma JEDEN markup pre vsetkych 5 mist') do
  NxTest.assert(NxStale::JS.include?('function refreshBtnHtml(stale, tip, attrs)'),
                'zdielany helper existuje')
  NxTest.assert_equal(1, NxStale::JS.scan(/id="refreshBtn"/).length,
                      'markup tlacidla je v celom studio.js PRAVE RAZ (5 kopii = 5 miest, ' \
                      'kde by sa jantar casom rozisiel)')
  # V budget.js smie `id="refreshBtn"` zit PRAVE RAZ — v NUDZOVEJ vetve mosta
  # (review #4): ked by helper zo studio.js nebol dostupny (parse chyba, zle
  # poradie skriptov), sekcie Rozpocet a Ponuka nesmu prist o JEDINU cestu
  # k cerstvym cislam. Lista sekcie ho nikdy nekresli priamo.
  NxTest.assert_equal(1, NxStale::BUD.scan(/id="refreshBtn"/).length,
                      'budget.js si vlastnu kopiu tlacidla nekresli — okrem nudzovej vetvy')
  bridge = NxStale::BUD[/function budRefreshBtnHtml\(tip\)\{.*?\n  \}/m].to_s
  NxTest.assert(bridge.include?('id="refreshBtn"'),
                'a ta jedna kopia je PRAVE v nudzovej vetve mosta')
  NxTest.refute(bridge.include?('nxstale'),
                'nudzove tlacidlo je NEUTRALNE — bez helpera nie je z coho jantar odvodit')
  # 3 volania v studio.js (Kusovnik, Kontrola, Nakup) + 2 v budget.js
  # (Rozpocet, Ponuka) = 5 mist, jeden markup.
  # `-1` = samotna definicia funkcie; zvysok su volacie miesta.
  NxTest.assert_equal(3, NxStale::JS.scan(/refreshBtnHtml\(/).length - 1,
                      'studio.js kresli tlacidlo v troch sekciach')
  NxTest.assert_equal(2, NxStale::BUD.scan(/budRefreshBtnHtml\('/).length,
                      'budget.js v dvoch (Rozpocet + Ponuka)')
  NxTest.assert_equal(1, NxStale::JS.scan(/t\.closest\('#refreshBtn'\)/).length,
                      'a vsetky idu JEDNYM handlerom (ziadna druha serverova cesta)')
end

NxTest.test('STALE: tooltip PRIZNAVA sirku signalu') do
  NxTest.assert(NxStale::JS.include?('V modeli nastali zmeny od posledného prepočtu'),
                'tooltip povie, PRECO je tlacidlo jantarove')
  NxTest.assert(NxStale::JS.include?('Platí pre akúkoľvek zmenu v dokumente, aj mimo skriniek.'),
                'a poctivo prizna, ze signal je sirsi nez kusovnik ' \
                '(posun cudzieho objektu cisla nezmeni)')
end

NxTest.test('STALE: stav zhadzuje VYHRADNE plny payload, echa nie') do
  set = NxStale::JS[/setStudio: function\(data\)\{.*?\n    \},/m].to_s
  NxTest.refute(set.empty?, 'setStudio sa nasiel')
  NxTest.assert(set.include?('staleFlag = false'), 'plny payload = cerstve cisla, jantar padá')
  NxTest.assert(set.index('staleFlag = false') < set.index('render()'),
                'a to PRED renderom — inak by lista este raz nakreslila jantar')
  # Echa nesu CISLA (lista VEPO, prepinace hran/kresby, vysledok zapisu
  # rozpoctu) — nesmu teda tvrdit, ze okno je aktualne.
  %w[setVepoBar setEdgeCheck setGrainCheck].each do |echo|
    body = NxStale::JS[/#{echo}: function\([^)]*\)\{.*?\n    \},/m].to_s
    NxTest.refute(body.include?('staleFlag'), "echo #{echo} stav NEZHADZUJE")
  end
  NxTest.refute(NxStale::BUD[/budgetResult: function.*?\n    \}/m].to_s.include?('staleFlag'),
                'ani echo vysledku zapisu rozpoctu')
  mark = NxStale::JS[/markStale: function\(\)\{.*?\n    \},/m].to_s
  NxTest.assert(mark.include?('renderTools()'), 'markStale prekresli LEN listu sekcie')
  NxTest.refute(mark.include?('renderBody') || mark.include?('render()'),
                'zoznam ani tabulka sa nedotykaju — su to stale tie iste cisla')
end

NxTest.test('STALE: jantar ide cez --nx-warn tokeny a ZIADNA zelena') do
  css = NxStale::HTML[/\.sectools \.ghostbtn\.bstalebtn,[^{]*\{[^}]*\}/m].to_s
  NxTest.assert(css.include?('.nxstale'), 'jantarove „Obnoviť" ma pravidlo')
  NxTest.assert(css.include?('--nx-warn'), 'farba ide vyhradne cez jantarove tokeny')
  NxTest.refute(css.match?(/green|--nx-ok|#[0-9a-fA-F]{6}/),
                'ziadna zelena ani vlastna farba — vyznamove farby ostavaju semaforu Kontroly')
end

File: tests/js/test_stale_obnovit.js
// Jantarovy indikator neaktualnosti tlacidla „Obnoviť" — KLIENTSKA cast (22.8.).
//
// Preco su to testy a nie klikanie:
//   1. Tlacidlo je na PIATICH miestach (Kusovnik · Kontrola · Nakup · Rozpocet ·
//      Ponuka). Keby si kazde kreslilo vlastny markup, jedna sekcia by casom
//      tvrdila, ze cisla su cerstve, hoci uz nie su — a prave z nej by sa
//      exportovalo. Sada preto overuje, ze vsetkych 5 ide JEDNYM helperom.
//   2. Stav sa smie zhodit VYHRADNE prichodom plneho payloadu. Echa (lista
//      VEPO, prepinace hran, vysledok zapisu rozpoctu) cisla NENESU — keby
//      jantar zhadzovali, okno by po zapise sumy tvrdilo, ze je aktualne.
//   3. `markStale` sa smie dotknut LEN listy sekcie: tabulka aj zoznam nalezov
//      su stale tie cisla, ktore sem prisli naposledy.
'use strict';
const assert = require('node:assert');
const path = require('node:path');

global.window = {};
const LISTEN = {};
const ELS = {};
global.document = {
  activeElement: null,
  addEventListener: function(type, fn){ (LISTEN[type] || (LISTEN[type] = [])).push(fn); },
  getElementById: function(id){ return ELS[id] || null; }
};
const S = require(path.join(__dirname, '..', '..', 'noxun_engine', 'ui', 'js', 'studio.js'));
// V prehliadaci JE `window` globalny objekt — v Node treba mat `NX` na oboch
// menach, inak by obal `NX.setStudio` v budget.js nenasiel, co obaluje.
// Poradie requirov je to iste ako poradie <script> v studio.html.
global.NX = global.window.NX;
const B = require(path.join(__dirname, '..', '..', 'noxun_engine', 'ui', 'js', 'budget.js'));

let n = 0;
function eq(a, b, msg){ n++; assert.deepStrictEqual(a, b, msg); }
function ok(c, msg){ n++; assert.ok(c, msg); }

const TIP_BOM = 'Prepočítať kusovník z aktuálneho modelu';

// --- 1) zdielany markup: dva stavy jedneho tlacidla --------------------------
(function(){
  const cold = S.refreshBtnHtml(false, TIP_BOM);
  const warm = S.refreshBtnHtml(true, TIP_BOM);
  ok(cold.indexOf('id="refreshBtn"') > -1, 'tlacidlo drzi svoje ID (zdielany handler)');
  ok(cold.indexOf('nxstale') < 0, 'aktualne okno = NEUTRALNE tlacidlo (ziadna trvala zelena)');
  ok(cold.indexOf(TIP_BOM) > -1, 'a tooltip hovori o TEJ sekcii');
  ok(warm.indexOf('class="ghostbtn nxstale"') > -1, 'neaktualne okno = jantarova trieda');
  ok(warm.indexOf(S.STALE_TIP) > -1, 'a tooltip povie PRECO');
  ok(warm.indexOf(TIP_BOM) > -1, 'bez toho, aby zahodil povodny popis sekcie');
  ok(warm.indexOf('Platí pre akúkoľvek zmenu v dokumente') > -1,
     'sirka signalu je PRIZNANA — jantar znamena „mozno neaktualne", nie „urcite zmenene"');
  // Lista Rozpoctu a Ponuky sa prekresluje aj pocas ovladania — bez `data-bkey`
  // by po prekresleni spadol fokus na <body>.
  ok(S.refreshBtnHtml(false, 'x', ' data-bkey="refresh"').indexOf('data-bkey="refresh"') > -1,
     'volajuci si smie doniest vlastne atributy (fokus listy Rozpoctu)');
  // Stav chodi ARGUMENTOM aj do listy Kusovnika (funkcia ostava cista).
  ok(S.bomToolsHtml({}, { view: 'parts', q: '', stale: true }).indexOf('nxstale') > -1,
     'lista Kusovnika kresli jantar z `st.stale`');
  ok(S.bomToolsHtml({}, { view: 'parts', q: '', stale: false }).indexOf('nxstale') < 0,
     'a bez neho ostava neutralna');
})();

// --- 2) rovnaky markup v Rozpocte aj Ponuke (budget.js je len most) ---------
(function(){
  eq(B.budStaleFlag(), false, 'bez signalu je okno povazovane za aktualne');
  const cold = B.budOfferToolsHtml();
  ok(cold.indexOf('id="refreshBtn"') > -1 && cold.indexOf('nxstale') < 0,
     'lista Ponuky ma „Obnoviť" a je neutralna');
  global.staleFlag = true;              // v prehliadaci je to globál studio.js
  eq(B.budStaleFlag(), true, 'budget.js cita stav z toho isteho miesta ako `ST`');
  const warm = B.budOfferToolsHtml();
  ok(warm.indexOf('nxstale') > -1, 'lista Ponuky zozltne s celym oknom');
  ok(warm.indexOf(S.STALE_TIP) > -1, 'a nesie TEN ISTY tooltip (jeden markup, jedna pravda)');
  ok(B.budToolsHtml({}).indexOf('nxstale') > -1, 'lista Rozpoctu tiez');
  eq(B.budRefreshBtnHtml('Prepočítať rozpočet z aktuálneho modelu'),
     S.refreshBtnHtml(true, 'Prepočítať rozpočet z aktuálneho modelu', ' data-bkey="refresh"'),
     'budget.js NEMA vlastnu kopiu markupu — vracia presne to, co studio.js');
  global.staleFlag = false;
})();

// --- 3) zivot stavu v okne: signal -> lista, payload -> zhodenie ------------
(function(){
  const tools = { innerHTML: '' };
  const body = { innerHTML: '' };
  ['snav', 'sechead', 'sectools', 'secbody', 'status', 'stModel'].forEach(function(id){
    ELS[id] = (id === 'sectools') ? tools : (id === 'secbody' ? body : { innerHTML: '', textContent: '' });
  });
  const SU = { ready: function(){}, refresh_bom: function(){} };
  global.window.sketchup = SU;
  global.sketchup = SU;
  const VEPO = { project: 'ENGINEtests', default_project: 'ENGINEtests', merge_18_36: true };
  const push = function(sec, gen){
    global.window.NX.setStudio({ version: '0.7.43', gen: gen, model_title: 'ENGINEtests',
                                 rows: [], sheets: [], counts: {}, control: [], vepo: VEPO,
                                 edge_check: { available: true, active: false, options: {}, counts: {} },
                                 grain_check: { available: true, active: false },
                                 open_section: sec });
  };

  push('bom', 1);
  ok(tools.innerHTML.indexOf('nxstale') < 0, 'po prepocte je lista Kusovnika neutralna');
  const bodyBefore = body.innerHTML;

  global.window.NX.markStale();
  ok(tools.innerHTML.indexOf('nxstale') > -1, 'signal zo servera zozltil „Obnoviť"');
  eq(body.innerHTML, bodyBefore,
     'a NEDOTKOL sa obsahu sekcie — su to stale tie cisla, ktore prisli naposledy');

  // Echa: nesu stav prepinacov a listy, NIE cisla.
  global.window.NX.setVepoBar({ project: 'Iná zákazka', merge_18_36: false });
  ok(tools.innerHTML.indexOf('nxstale') > -1, 'echo listy VEPO jantar NEZHADZUJE');

  push('ctrl', 2);
  ok(tools.innerHTML.indexOf('nxstale') < 0, 'plny payload ho zhodil aj v Kontrole');
  global.window.NX.markStale();
  ok(tools.innerHTML.indexOf('nxstale') > -1, 'Kontrola zozltne rovnako ako Kusovnik');
  ok(tools.innerHTML.indexOf('Prepočítať kontrolu z aktuálneho modelu') > -1,
     'a drzi si vlastny popis sekcie');
  global.window.NX.setEdgeCheck({ available: true, active: true, options: {}, counts: {} });
  ok(tools.innerHTML.indexOf('nxstale') > -1,
     'echo prepinaca hran prekresli listu, ale jantar NEZHADZUJE (nenesie cisla)');
  global.window.NX.setGrainCheck({ available: true, active: true });
  ok(tools.innerHTML.indexOf('nxstale') > -1, 'ani echo smeru kresby');

  // Opakovany signal je no-op (server ho posiela raz za burst, ale okno sa
  // nesmie prekreslovat pri kazdom dalsom).
  global.window.NX.markStale();
  ok(tools.innerHTML.indexOf('nxstale') > -1, 'druhy signal stav nemeni');

  push('buy', 3);
  ok(tools.innerHTML.indexOf('nxstale') < 0, 'a Nakup kovania sa sprava rovnako');
  global.window.NX.markStale();
  ok(tools.innerHTML.indexOf('nxstale') > -1, 'jantar sa ukaze aj v Nakupe');
})();

console.log(`test_stale_obnovit: ${n} kontrol OK`);
